classes/observation_manager.php

```diff
@@ -0,0 +1,402 @@
+     * @return array array of database objects obtained from database
+     */
+    public static function get_observation_points(int $observationid, string $sortby='list_order',
+            string $tablename='observation_points'): array {
+        global $DB;
+        return $DB->get_records($tablename, ['obs_id' => $observationid], $sortby);
+    }
+
+    /**
+     * Deletes observation point
+     * @param int $observationid ID of the observation instance
+     * @param int $obpointid ID of the observation point to delete
+     * @param string $tablename database table name
+     */
+    public static function delete_observation_point(int $observationid, int $obpointid, string $tablename='observation_points') {
+        global $DB;
+        // To ensure ordering stays intact, should move all those points with a higher order than this one down by 1.
+        $currentpoint = self::get_existing_point_data($observationid, $obpointid, $tablename);
+
+        // Get those with a higher list ordering than this one.
+        $pointsabove = $DB->get_records_select(
+            $tablename,
+            "obs_id = :obsid AND list_order > :listorder",
+            [
+                'obsid' => $observationid,
+                'listorder' => $currentpoint->list_order
+            ]
+        );
+
+        $transaction = $DB->start_delegated_transaction();
+
+        $DB->delete_records($tablename, ['id' => $obpointid, 'obs_id' => $observationid]);
+
+        // Shuffle those above down.
+        foreach ($pointsabove as $pointabove) {
+            $DB->update_record(
+                $tablename,
+                [
+                    'id' => $pointabove->id,
+                    'list_order' => $pointabove->list_order - 1
+                ]
+            );
+        }
+
+        $transaction->allow_commit();
+    }
+
+    /**
+     * Reorders an observation point in relation to the other observation points in this observation instance
+     * @param int $observationid ID of the observation instance
+     * @param int $obpointid ID of the observation point to reorder
+     * @param int $direction direction and magnitude to reorder the point in
+     * @param string $tablename database table name
+     */
+    public static function reorder_observation_point(int $observationid, int $obpointid, int $direction,
+            string $tablename='observation_points') {
+        global $DB;
+
+        if (!is_int($direction) || $direction == 0) {
+            throw new \coding_exception("direction must be an integer that is not zero");
+        }
+
+        // First get the ordering of the current point.
+        $targetpoint = self::get_existing_point_data($observationid, $obpointid, $tablename);
+
+        // Get all the points to get the bounds.
+        $allpoints = self::get_observation_points($observationid);
+        $alllistorders = array_column($allpoints, 'list_order');
+
+        if (count($alllistorders) === 0) {
```

Comment:
Can this ever happen? I would expect the field to not be nullable in the DB, or this check should perhaps be for the row count overall